pkcs7/pkcs7.go

```diff
@@ -0,0 +1,145 @@
+	oidPKCS7Data          = asn1.ObjectIdentifier{1, 2, 840, 113549, 1, 7, 1}
+	oidPKCS7SignedData    = asn1.ObjectIdentifier{1, 2, 840, 113549, 1, 7, 2}
+	oidPKCS9ContentType   = asn1.ObjectIdentifier{1, 2, 840, 113549, 1, 9, 3}
+	oidPKCS9MessageDigest = asn1.ObjectIdentifier{1, 2, 840, 113549, 1, 9, 4}
+	oidPKCS9SigningTime   = asn1.ObjectIdentifier{1, 2, 840, 113549, 1, 9, 5}
+	oidSHA1               = asn1.ObjectIdentifier{1, 3, 14, 3, 2, 26}
+)
+
+// Sign data with cert & private key.
+func Sign(data io.Reader, cert *x509.Certificate, priv *rsa.PrivateKey, intermediate *x509.Certificate) ([]byte, error) {
+	var hash = sha1.New()
+	if _, err := io.Copy(hash, data); err != nil {
+		return nil, err
+	}
+
+	raw := cert.Raw
+	// Passbook needs Apple's intermediate WWDR certificate.
+	if intermediate != nil {
+		raw = append(raw, intermediate.Raw...)
+	}
+
+	var signedData = signedData{
+		Version: 1,
+		DigestAlgorithms: []algorithmIdentifier{{
+			Algorithm:  oidSHA1,
+			Parameters: asn1.RawValue{Tag: 5},
+		}},
+		ContentInfo: contentInfo{Type: oidPKCS7Data},
+		Certificates: asn1.RawValue{
+			Class: 2, Tag: 0, Bytes: raw, IsCompound: true,
+		},
+		SignerInfos: []signerInfo{{
+			Version: 1,
+			IssuerAndSerialNumber: issuerAndSerialNumber{
+				Issuer:       asn1.RawValue{FullBytes: cert.RawIssuer},
+				SerialNumber: cert.SerialNumber,
+			},
+			DigestAlgorithm: algorithmIdentifier{
+				Algorithm:  oidSHA1,
+				Parameters: asn1.RawValue{Tag: 5},
+			},
+			AuthenticatedAttributes: []attribute{
+				newAttribute(oidPKCS9ContentType, oidPKCS7Data),
+				newAttribute(oidPKCS9SigningTime, time.Now().UTC()),
+				newAttribute(oidPKCS9MessageDigest, hash.Sum(nil)),
+			},
+			DigestEncryptionAlgorithm: algorithmIdentifier{
+				Algorithm:  oidPKCS1RSAEncryption,
+				Parameters: asn1.RawValue{Tag: 5},
+			},
+		}},
+	}
+	encodedAuthenticatedAttributes, err := asn1.Marshal(
+		signedData.SignerInfos[0].AuthenticatedAttributes)
+	if err != nil {
+		return nil, err
+	}
+	// For the digest of the authenticated attributes, we need a
+	// slightly different encoding. Change the attributes from a
+	// SEQUENCE to a SET.
+	var originalFirstByte = encodedAuthenticatedAttributes[0]
+	encodedAuthenticatedAttributes[0] = 0x31
+	hash = sha1.New()
+	hash.Write(encodedAuthenticatedAttributes)
+	var attributesDigest = hash.Sum(nil)
+	encodedAuthenticatedAttributes[0] = originalFirstByte
+	encryptedDigest, err := rsa.SignPKCS1v15(rand.Reader, priv, crypto.SHA1, attributesDigest)
+	if err != nil {
+		return nil, err
+	}
+	signedData.SignerInfos[0].EncryptedDigest = encryptedDigest
+	return asn1.Marshal(container{
+		OID:        oidPKCS7SignedData,
+		SignedData: signedData,
+	})
+}
+
+type container struct {
+	OID        asn1.ObjectIdentifier
+	SignedData signedData `asn1:"tag:0,explicit,optional"`
+}
+
+// signedData is defined in rfc2315, section 9.1.
+type signedData struct {
+	Version          int                   `asn:"default:1"`
```

Comment:
CFSSL's [ParsePKCS7](https://github.com/cloudflare/cfssl/blob/master/crypto/pkcs7/pkcs7.go) gives the same error.

I'm not sure if I'm doing this right, but calling openssl to verify the output gives this error:

```bash
> /usr/local/opt/openssl/bin/openssl smime -verify -in signature-test
Error reading S/MIME message
140735153643600:error:0D0D40D1:asn1 encoding routines:SMIME_read_ASN1:no content type:asn_mime.c:440:
```